mountpoint-s3/src/fs.rs

```diff
@@ -726,10 +698,29 @@ where
             return Err(err!(libc::EINVAL, "O_SYNC and O_DSYNC are not supported"));
         }
 
-        // All file handles will be lazy initialized on first read/write.
-        let state = FileHandleState::new(lookup, flags, pid).await.into();
+        let state = if flags & libc::O_RDWR != 0 {
+            let is_truncate = flags & libc::O_TRUNC != 0;
+            if !remote_file || (self.config.allow_overwrite && is_truncate) {
+                // If the file is new or opened in truncate mode, we know it must be a write handle.
+                debug!("fs:open choosing write handle for O_RDWR");
+                FileHandleState::new_write_handle(&lookup, lookup.inode.ino(), flags, pid, self).await?
+            } else {
+                // Otherwise, it must be a read handle.
+                debug!("fs:open choosing read handle for O_RDWR");
+                FileHandleState::new_read_handle(&lookup, self).await?
+            }
+        } else if flags & libc::O_WRONLY != 0 {
+            FileHandleState::new_write_handle(&lookup, lookup.inode.ino(), flags, pid, self).await?
+        } else {
+            FileHandleState::new_read_handle(&lookup, self).await?
```

Comment:
does prefetching starts on `open` or on `read` with this change? when prefetching starts in current MP version?

Reply:
We now initialize the prefetcher on open, but we still only spawn requests on the first read.

Reply:
that resolves my concern 👍